hydra/_internal/hydra.py

```diff
@@ -103,15 +104,23 @@ def run(
             run_mode=RunMode.RUN,
         )
         HydraConfig.instance().set_config(cfg)
+        callbacks = get_callbacks(cfg)
 
-        return run_job(
+        for c in callbacks:
+            c.pre_run(cfg)
+        ret = run_job(
             config=cfg,
             task_function=task_function,
             job_dir_key="hydra.run.dir",
             job_subdir_key=None,
             configure_logging=with_log_configuration,
         )
 
+        for c in reversed(callbacks):
+            c.post_run(cfg)
```

Comment:
same exception handling logic.
this will repeat so you should probably introduce a function to actually loop and call the callbacks.